### ISSUE TITLE
Dynamiske uttrykk: Mulighet til å vise tekst fra skjult felt

### ISSUE THREAD

**Issue by StianVestli** [Altinn/app-frontend-react#1669]
### Description

Man har i dag muligheten til å hente ut visningstekster fra komponenter via dynamiske uttrykk:
`["displayValue", "component-id"]`
Vi bruker en del prefill data som vi vil presentere for brukeren gjerne da som tekst. For å få denne teksten formatert så vil vi kunne sette denne til et inputfelt og så legge dette skjult for brukeren. Problemet i dag er da at vi ikke kan hente ut dislayValue.

Vi trenger å kunne hente ut : `["hiddenDisplayValue", "component-id"]`

### Avhengig av Altinn/app-lib-dotnet#1157
~~Denne oppgaven kan ansees å være avhengig av Altinn/app-lib-dotnet#1157. Når vi uansett "fjerner data" for skjulte komponenter vil uansett oppslag av `displayValue` (og `component`-oppslag) påvirkes av dette. Det som da blir riktig oppførsel er jo å vise skjult data på samme måte som resten av skjemamotoren gjør - for selv om komponenten er skjult, kan den ha en verdi (f.eks. preutfyllt ved oppretting av datamodellen) som vi ikke skal slette (for vi skal i utgangspunktet bare slette data brukeren har fyllt ut, og det vil også si at vi må "rulle tilbake" til slik datamodellen var _før brukeren gjorde endringer_ når brukeren har gjort slik at komponentene ble skjult).~~

~~Med andre ord, opter man inn for å bruke funksjonaliteten i Altinn/app-lib-dotnet#1157 er det også naturlig at vi fjerner denne begrensningen som gjør at `displayValue` og `component` funksjonene i uttrykk gir tom verdi om man spør etter skjulte komponenter.~~

**Oppdatering, 10.01.2025**: Nå ser Altinn/app-lib-dotnet#1157 ut til å dreie i en litt annen retning (da dette nok heller blir implementert på backend), og da blir det ikke hensiktsmessig å gjøre seg avhengig av en tenkt løsning der (når uttrykkene må kunne kjøre i frontend).

### Additional Information

- Altinn/app-lib-dotnet#1157
- #2287

**Comment by olemartinorg** [comment]
Vi diskuterte denne på statusmøtet i dag 20. mars, og det kom frem at behovet her kanskje i like stor grad er å kunne tall-formattere noe man henter fra datamodellen direkte. Men dette gjelder jo også f.eks. option-lister, og da måtte man kunne hatt støtte for å vise frem en label for en gitt option-verdi.

Så selve behovet her er gjerne å kunne vise slik tekst, via uttrykk, med eller uten å konfigurere opp en skjult komponent for å beskrive hvordan man skal formattere visningsdataene.

**Comment by olemartinorg** [comment]
Diskutert i møte. Denne har et visst overlapp med `optionLabel`-funksjonen (beskrevet i #1299) og en mulighet for å bruke tallformattering (og datoformattering?) i uttrykk.

**Comment by StianVestli** [comment]
Hvis vi sender inn `optionId` , slik `["optionLabel", "code-list-id", "some-option-value"]` så er det vel slik at vi sjølv må si hvilken verdi vi vil ha ut av lista. Det er mulig det er en god løsning. Men saken ble opprinnelig skrevet med tanke på at man kan bruke et felt for å få ut formateringer og feks prefix, men også dette med dropdown som  ligger i en repeterende gruppe, og at det var en verdi ut fra denne vi vil ha som da er hentet utfra koblingen den har mot verdi som ligger i gruppa. Så dette må også være mulig hvis en skal linke inn i uttrykket `["optionLabel", "code-list-id", "some-option-value"]`. Dette med å kunne se tekst fra skjulte felter var jo også for å få formatering av feks tall, men ser ut som du har pekt på en annen sak som vil løse det. Litt usikker på hva du mener her :"Begrensende om man bruker options fra f.eks. repgruppe eller en statisk liste." hvis det betyr at man ikke kan bruke dette i repgruppe eller med statiske optionlister (JSON) så vil vel det å gå mot komponent være bedre. Vi vil jo og bruke dette i 2 nivåer i gruppe i gruppe, slik at en kan referere til verdi i en lister i gruppe1 i en tekst som ligger i gruppe2.

**Comment by olemartinorg** [comment]
Takk for tilbakemelding, @StianVestli! 🙌

> Men saken ble opprinnelig skrevet med tanke på at man kan bruke et felt for å få ut formateringer og feks prefix, men også dette med dropdown som ligger i en repeterende gruppe, og at det var en verdi ut fra denne vi vil ha som da er hentet utfra koblingen den har mot verdi som ligger i gruppa.

Alt dette burde være mulig slik jeg har sett for meg `optionLabel`-funksjonen, ja. 😊

> Litt usikker på hva du mener her :"Begrensende om man bruker options fra f.eks. repgruppe eller en statisk liste." hvis det betyr at man ikke kan bruke dette i repgruppe eller med statiske optionlister (JSON) så vil vel det å gå mot komponent være bedre.

Håper den [nylig oppdaterte dokumentasjonen rundt dette](https://docs.altinn.studio/nb/altinn-studio/guides/development/options/sources/) kan være oppklarende. Det finnes jo 3 forskjellige egenskaper vi kan bruke til å konstruere en kodeliste, nemlig `options`, `optionsId` og `source`. Så hvis vi lager funksjonen slik at den bare har støtte for `optionId` er den jo begrenset til statiske kodelister fra json-filer, dynamiske kodelister (fra backend) og delte kodelister (ref tabellen i docs).

Det gjør jo også at man mister:
- Støtte for helt statiske `options` definert på komponent-nivå
- Kodelister [fra repeterende grupper](https://docs.altinn.studio/nb/altinn-studio/guides/development/options/sources/from-data-model) - men jo, man kan fortsatt bruke `optionLabel`-funksjonen på komponenter _inne i repeterende grupper_.
- Diverse ålreite funksjonaliteter som [spørringsparametre](https://docs.altinn.studio/nb/altinn-studio/guides/development/options/sources/dynamic/#sp%C3%B8rringsparametre) og [filtrering](https://docs.altinn.studio/nb/altinn-studio/guides/development/options/functionality/filtering/)

Hvis man hadde knyttet dette til "alle options en gitt komponent har" får man jo i pose og sekk, men ulempen er at man må konfigurere opp denne komponenten om man ikke har den allerede, og hvis man ender opp med å konfigurere en komponent som alltid er skjult og som bare brukes som en kilde til options for `optionLabel`, har det kanskje gått litt for langt.

**Comment by StianVestli** [comment]
Det er vel den siste setningen din som kanskje gir svaret når je tenker meg om ,at det å gå mot optionid og ikke komponent som er riktige veien å gå. Jeg sliter litt med å se alle senarioer her for å være ærlig, men å "misbruke" en komponent bare for å få til dette blir kanskje feil.

**Comment by olemartinorg** [comment]
Flott, da tenker vi likt! 🙏 Jeg tenker vi går for `optionsId`, og dermed regner jeg oppgaven som klar til å begynne på! Takk!

**Comment by olemartinorg** [comment]
@StianVestli Da er `optionLabel`-funksjonen ute. 🙌 Når du får anledning til å teste den ut, gir du meg et lite vink om den funker som tenkt? 😊 Docs her: https://docs.altinn.studio/nb/altinn-studio/reference/logic/expressions/#func-optionLabel

**Comment by StianVestli** [comment]
Skal si ifra når jeg har et skjema som dette trengs på igjen. men ser i docen at det står den krever 2 parametere men den viser 3 i docen:

![Image](https://github.com/user-attachments/assets/6631995b-20bc-4d50-a9c3-b7ceb01c3cbf)

**Comment by olemartinorg** [comment]
Ja, første verdi er alltid funksjonsnavnet. Så funksjonsnavnet + 2 argumenter = 3 verdier i arrayen.

**Comment by StianVestli** [comment]
stemmer , selvfølgelig... veit ikke hvorfor je trudde det var optionsnavnet, hang ikke helt med tydeligvis :)

**Comment by StianVestli** [comment]
Testet nå og det ser ut som det funker det på det lille je testa nå.

**Comment by olemartinorg** [comment]
Da lukker jeg denne saken da det meste er ferdig her, og skiller ut gjenværende del til #3146 (altså uttrykksfunksjon for formattering av tall). Det ble litt mye å lese seg opp på i denne issuen om man skulle finne frem til essensen av det som faktisk var igjen.

**Comment by rvessb** [comment]
@olemartinorg ser greit ut det. Et lite oppfølgingsspørsmål, for å utelukke at vi faktisk trenger en "showhiddenvalue" fuksjon...
Dersom jeg trenger en verdi fra en kodeliste feks 5 ganger på en side: Er det noen forskjell, slik ressursmessig, om jeg bruker uttrykket for `optionLabel` og kaller kodelista 5 ganger (er en .cs-liste som knar litt på flerer kodelsiter fra KLASS) fremfor å gjøre kallet en gang, legge det som `Value` på en Text-komponent, og peke på denne `Text`-komonenten de 4 andre gangene?

**Comment by olemartinorg** [comment]
> Er det noen forskjell, slik ressursmessig, om jeg bruker uttrykket for optionLabel og kaller kodelista 5 ganger (er en .cs-liste som knar litt på flerer kodelsiter fra KLASS) fremfor å gjøre kallet en gang, legge det som Value på en Text-komponent, og peke på denne Text-komonenten de 4 andre gangene?

Ja, foreløpig er sistnevnte (legge til en unødvendig `Text`-komponent) litt mer ressurskrevende. På sikt skal vi nok dit at det blir tilnærmet likt, men jeg tror nok alltid at en ekstra komponent vil kreve bittelitt mer prosessering på klientsiden fremfor å bare bruke kodelisten direkte.
